packages/dd-trace/test/appsec/waf/index.spec.js

```diff
@@ -122,6 +139,100 @@ describe('WAF Manager', () => {
+      })
+
+      it('should call updateRateLimitedMetric when result.keep is true but rate limiter denies', () => {
+        const result = { keep: true, events: [] }
+        mockWafContext.run.returns(result)
+        limiterStub.isAllowed.returns(false)
+
+        const payload = { persistent: { 'server.io.net.url': 'http://example.com' } }
+        waf.run(payload, req)
+
+        expect(updateRateLimitedMetric).to.have.been.calledOnceWithExactly(req)
+        expect(keepTrace).not.to.have.been.called
+      })
+
+      it('should not call keepTrace or updateRateLimitedMetric when result.keep is false', () => {
+        const result = { keep: false, events: [] }
+        mockWafContext.run.returns(result)
+        limiterStub.isAllowed.returns(true)
+
+        const payload = { persistent: { 'server.io.net.url': 'http://example.com' } }
+        waf.run(payload, req)
+
+        expect(keepTrace).not.to.have.been.called
+        expect(updateRateLimitedMetric).not.to.have.been.called
+      })
+
+      it('should not call keepTrace or updateRateLimitedMetric when result.keep is undefined', () => {
+        const result = { events: [] }
+        mockWafContext.run.returns(result)
+        limiterStub.isAllowed.returns(true)
+
+        const payload = { persistent: { 'server.io.net.url': 'http://example.com' } }
+        waf.run(payload, req)
+
+        expect(keepTrace).not.to.have.been.called
+        expect(updateRateLimitedMetric).not.to.have.been.called
+      })
+
+      it('should not call keepTrace or updateRateLimitedMetric when result is null', () => {
+        mockWafContext.run.returns(null)
+        limiterStub.isAllowed.returns(true)
+
+        const payload = { persistent: { 'server.io.net.url': 'http://example.com' } }
+        waf.run(payload, req)
+
+        expect(keepTrace).not.to.have.been.called
+        expect(updateRateLimitedMetric).not.to.have.been.called
+      })
+
+      it('should handle keep=true with events and rate limiter allowing', () => {
+        const result = { keep: true, events: ['attack_event'] }
+        mockWafContext.run.returns(result)
+        limiterStub.isAllowed.returns(true)
+
+        const payload = { persistent: { 'server.io.net.url': 'http://example.com' } }
+        waf.run(payload, req)
+
+        expect(keepTrace).to.have.been.calledOnceWithExactly({ mock: 'rootSpan' }, 'ASM')
+        expect(updateRateLimitedMetric).not.to.have.been.called
+      })
+
+      it('should handle keep=true with events but rate limiter denying', () => {
+        const result = { keep: true, events: ['attack_event'] }
+        mockWafContext.run.returns(result)
+        limiterStub.isAllowed.returns(false)
+
+        const payload = { persistent: { 'server.io.net.url': 'http://example.com' } }
+        waf.run(payload, req)
+
+        expect(updateRateLimitedMetric).to.have.been.calledOnceWithExactly(req)
+        expect(keepTrace).not.to.have.been.called
```

Comment:
those two tests are very similar to the ones above, the only difference is the presence of events ? is that really neccesary to have specific tests for these cases ?